### PR TITLE
verbs: make sure the memory region to send from is registered

### LINKED ISSUE #23220: verbs: local protection error when doing rdma send
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 14.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 1.9.0
- Python version: 2.7
- Bazel version (if compiling from source): 0.16.1
- GCC/Compiler version (if compiling from source): 6.3
- CUDA/cuDNN version: 9.0
- GPU model and memory: TITAN Xp. 12 GB

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

**Describe the current behavior**
When verbs support is enabled, getting

```
mlx5: 4dfe187391fa: got completion with error:
00000000 00000000 00000000 00000000
00000000 00000000 00000000 00000000
00000000 00000000 00000000 00000000
00000000 92005204 0900013b 05cc37d3
2018-09-21 23:48:04.608288: F external/org_tensorflow/tensorflow/contrib/verbs/rdma.cc:451] Check failed: wc_[i].status == IBV_WC_SUCCESS Failed status 
local protection error 4 139821989727968 82
```

**Describe the expected behavior**
Should not produce this error.

**Code to reproduce the issue**
Need substantial infrastructure (RoCE) and code to reproduce.

**Other info / logs**
The `wr`s (verbs work request) that trigger this error has lkey set to 0 and non-zero buffer length, indicating that it's trying to send unregistered memory region.

### PR DESCRIPTION
**Pull request by guoshimin**
Fixes #23220

### CODE REVIEW

**Comment by xiejw** [comment]
I am not the expert for this. Re: Admin, please find a new reviewer.

**Comment by yanivbl6** [comment]
Thanks for the PR. Did you test it on 1.9?
I am asking because verbs is currently fails on build. 

When testing it on my own branch, it did seem to fix #23220, but I am worried about the registered memory not being properly de-registered.

**Comment by guoshimin** [comment]
Yes I tested it on 1.9. Does the mr get deregistered in `RdmaTensorResponse::Destroy()`?

**Comment by yanivbl6** [comment]
It is normally dereged from MRDestroy, called on iterator erasure 
`void RdmaMemoryMgr::EvictMemoryRegion(void* addr, size_t length);`

You can use InsertMemoryRegion to register memory so it can also be used in the future (using FindMemoryRegion) and automatically evicted in the end.

**Comment by byronyi** [comment]
The cost of registration on every single transmission is higher than cost of actual transmission for small sized tensor (could be up to 32KB). 

I believe @yanivbl6 is working on a better fix.

**Comment by yanivbl6** [comment]
If you use InsertMemoryRegion, next attempts to access the same memory should work as with FindMemoryRegion without extra registration, so performance will be fine in the long run as long it's not just random data access (unlikely in the case of graph-computation).

I don't want people to wait for my fix as I am having an hard time finding the time to work on this.
that being said, I would like to have this tested on master due to the major changes pushed to the memory allocation mechanism since 1.9 (In the same commit that also broke the build).

**Comment by poxvoculi** [comment]
I defer to @yanivbl6 on any changes to contrib/verbs.  When he's ready to accept, I will approve.  (I can't test any of this code.)

**Comment by guoshimin** [comment]
@yanivbl6 Thanks for your explanation. Changed to use InsertMemoryRegion. Not sure what allocator_name should be set to. Seems it's only for display purposes.

**Review by yanivbl6 — CHANGES_REQUESTED** [review]

**Comment by yanivbl6** [comment]
Sorry, I added the comments but forgot to submit the review.

File: tensorflow/contrib/verbs/rdma.cc [hunk 1]
```diff
@@ -1204,6 +1204,12 @@ void RdmaTensorResponse::SendContent(const Tensor& in, const TensorProto& proto,
         src_addr_ = src_buffer_->data();
         mr_ = RdmaMemoryMgr::Singleton().FindMemoryRegion(src_addr_,
                                                           tensor_bytes);
+        if (mr_ == nullptr && tensor_bytes) {
+          RdmaMemoryMgr::Singleton().InsertMemoryRegion(
+              src_addr_, tensor_bytes, "what should go here?");
```
[the remark this answers is not in the document]
Comment (guoshimin) [reply tensorflow/contrib/verbs/rdma.cc]: Done

**Comment by yanivbl6** [comment]
OK with me.
noting again that verbs is still broken in master branch, so one would need to cherry-pick this patch from older version you get this fix.

**Comment by guoshimin** [comment]
@yanivbl6 Do you know what's the last release version where verbs still built?

@poxvoculi Can you approve? Thanks!

**Review by poxvoculi — DISMISSED** [review]

**Comment by yanivbl6** [comment]
The patch that broke the code was [33170cc)](https://github.com/tensorflow/tensorflow/commit/33170cc661f3838aa7d0d7fc19bb0c6ba4812a3c)

The last version I tried myself which worked was 1.9, but 1.10 might also work.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @wt-huang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by 2sin18** [comment]
I patched this PR in r1.12 branch with other commits cherry-picked from master branch, but "local protection error" still exists for slim based inception model.

**Comment by byronyi** [comment]
I believe the issue has been fixed in #24250 without this PR. @2sin18 @guoshimin could you confirm?

**Comment by 2sin18** [comment]
I have cherry-pick your commit from master, but it still not works in r1.12. I will try this patch in r1.13 branch. Thanks.

**Comment by 2sin18** [comment]
I add `force_gpu_compatible=True` to `GPUOptions`, and #24250 works without this PR.

**Comment by byronyi** [comment]
You could patch 7dea8383bbb97a1e78cdece876e083b22191974e as well.

**Comment by guoshimin** [comment]
I can't test the new fix since we are still using 1.9.  I'll close my PR for now.

**Pull request closed without merging**